File: app/src/lib/players/BuildNovaPredictPlayerIdentityCrosswalkKey.ts
/*
  BuildNovaPredictPlayerIdentityCrosswalkKey.ts
  ---------------------------------------------
  Deterministic join key for Sleeper catalog rows ↔ nflverse weekly actuals.

  Mirrors pipeline/lib/BuildNovaPredictPlayerIdentityCrosswalkKey.py so the web app
  can query player_weekly_actuals without maintaining a separate ID map table yet.
*/

export function BuildNovaPredictPlayerIdentityCrosswalkKey(
  fullName: string,
  team: string | null | undefined,
  position: string | null | undefined,
): string {
  const normalizedName = fullName.toLowerCase().split(/\s+/).filter(Boolean).join(" ");
  const normalizedTeam = (team ?? "").trim().toLowerCase();
  const normalizedPosition = (position ?? "").trim().toLowerCase();
  return `${normalizedName}|${normalizedTeam}|${normalizedPosition}`;
}

File: pipeline/lib/BuildNovaPredictPlayerIdentityCrosswalkKey.py
"""
BuildNovaPredictPlayerIdentityCrosswalkKey
--------------------------------------------
Builds a deterministic lowercase key for joining Sleeper catalog rows to nflverse stats.

Why name+team+position instead of ID maps:
- Sleeper `gsis_id` is often null (verified on CeeDee Lamb payload).
- nflverse uses GSIS-style IDs; Sleeper uses numeric IDs.
- Full name + team + position is stable enough for weekly stats joins at our scale,
  and both catalogs are refreshed from the same real NFL universe.

Called by:
- NormalizeNflverseWeeklyStatsIntoPlayerWeeklyActualsTable.py
- ComputeWeeklyProjectionsFromRealWeeklyStatsJob.py
"""


def BuildNovaPredictPlayerIdentityCrosswalkKey(
    full_name: str,
    team: str | None,
    position: str | None,
) -> str:
    """
    Returns a normalized join key like "ceedee lamb|dal|wr".
    """
    normalized_name = " ".join(full_name.lower().split())
    normalized_team = (team or "").strip().lower()
    normalized_position = (position or "").strip().lower()
    return f"{normalized_name}|{normalized_team}|{normalized_position}"
